src/libraries/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs

```diff
@@ -65,37 +73,55 @@ internal bool IsReadOnly(ReadOnlySpan<char> path, bool continueOnError = false)
             }
 #endif
 
-            return ((_fileStatus.Mode & (int)readBit) != 0 && // has read permission
-                (_fileStatus.Mode & (int)writeBit) == 0);     // but not write permission
+            return (fileStatus.Mode & (int)readBit) != 0 && // has read permission
+                (fileStatus.Mode & (int)writeBit) == 0;     // but not write permission
         }
 
-        public FileAttributes GetAttributes(ReadOnlySpan<char> path, ReadOnlySpan<char> fileName)
+        internal static bool IsSymLink(Interop.Sys.FileStatus fileStatus)
         {
-            // IMPORTANT: Attribute logic must match the logic in FileSystemEntry
-
-            EnsureStatInitialized(path);
-
-            if (!_exists)
-                return (FileAttributes)(-1);
+            return (fileStatus.Mode & Interop.Sys.FileTypes.S_IFMT) == Interop.Sys.FileTypes.S_IFLNK;
+        }
 
+        internal static FileAttributes GetAttributes(bool isReadOnly, bool isSymlink, bool isDirectory, bool isHidden)
+        {
             FileAttributes attributes = default;
 
-            if (IsReadOnly(path))
+            if (isReadOnly)
                 attributes |= FileAttributes.ReadOnly;
-
-            if ((_fileStatus.Mode & Interop.Sys.FileTypes.S_IFMT) == Interop.Sys.FileTypes.S_IFLNK)
+            if (isSymlink)
                 attributes |= FileAttributes.ReparsePoint;
-
-            if (_isDirectory)
+            if (isDirectory)
                 attributes |= FileAttributes.Directory;
-
-            // If the filename starts with a period or has UF_HIDDEN flag set, it's hidden.
-            if (fileName.Length > 0 && (fileName[0] == '.' || (_fileStatus.UserFlags & (uint)Interop.Sys.UserFlags.UF_HIDDEN) == (uint)Interop.Sys.UserFlags.UF_HIDDEN))
+            if (isHidden)
                 attributes |= FileAttributes.Hidden;
 
             return attributes != default ? attributes : FileAttributes.Normal;
         }
 
+        internal void Invalidate() => _fileStatusInitialized = -1;
+
+        internal bool IsReadOnly(ReadOnlySpan<char> path, bool continueOnError = false)
+        {
+            EnsureStatInitialized(path, continueOnError);
+            return IsReadOnly(_fileStatus);
+        }
+
+        public FileAttributes GetAttributes(ReadOnlySpan<char> path, ReadOnlySpan<char> fileName)
+        {
+            // IMPORTANT: Attribute logic must match the logic in FileSystemEntry
+
+            EnsureStatInitialized(path);
+
+            if (!_exists)
+                return (FileAttributes)(-1);
+
+            return GetAttributes(
+                IsReadOnly(path),
+                IsSymLink(_fileStatus),
+                _isDirectory,
+                (fileName.Length > 0 && fileName[0] == '.') || IsHidden(_fileStatus));
+        }
+
         public void SetAttributes(string path, FileAttributes attributes)
```

Comment:
As for throw, in line 131 below Hidden attribute presents as valid - I think we could exclude it on systems which do not support this. (I don't know how detect this in right way.)

Reply:
If `FileAttribute.Hidden` is not supported on Linux, we should update the  docs since they say otherwise https://docs.microsoft.com/dotnet/api/system.io.fileattributes?view=netcore-3.1

Reply:
I'm not convinced we have to throw for this.

If the flag is not supported in the current OS, then it will never be set, so any condition checking if the UF_HIDDEN flag was set, will be false.

Reply:
See line 147, we are already checking that. No need to throw.

```cs
            if (Interop.Sys.CanSetHiddenFlag)
            {
```